### PR TITLE
storage: listing noncurrent object versions

### PR DESCRIPTION
**Pull request by drakoleat**
sample code for listing versioned objects

### CODE REVIEW

**Comment by drakoleat** [comment]
CC @tbpg could you please take a look?

File: internal/testutil/storage_cleaner.go [hunk 1]
```diff
@@ -76,9 +79,16 @@ func deleteBucketIfExists(ctx context.Context, t *testing.T, client *storage.Cli
 				t.Errorf("Bucket(%q).Object(%q).Update: %v", bucket, attrs.Name, err)
 			}
 		}
+		if attrs.Generation != 0 {
+			if err := b.Object(attrs.Name).Generation(attrs.Generation).Delete(ctx); err != nil {
+				t.Errorf("Bucket(%q).Object(%q).Generation(%q).Delete: %v", bucket, attrs.Name, attrs.Generation, err)
+			}
+			continue
```
[the remark this answers is not in the document]
Comment (drakoleat) [reply internal/testutil/storage_cleaner.go:86]: Object deletion function call is different for bucket with versioning. `continue` to skip the delete function call below it. 

Would if..else.. statement make this easier to understand?
```
// object with version
if attrs.Generation != 0 {
    // delete with Generation()
} else {
    // normal delete
}
```

File: storage/objects/objects_test.go [hunk 2]
```diff
@@ -53,6 +54,16 @@ func TestObjects(t *testing.T) {
 
 	testutil.CleanBucket(ctx, t, tc.ProjectID, bucket)
 	testutil.CleanBucket(ctx, t, tc.ProjectID, dstBucket)
+	testutil.CleanBucket(ctx, t, tc.ProjectID, bucketVersioning)
+
+	{
+		// Enable versioning
+		_, err := client.Bucket(bucketVersioning).Update(ctx,
+			storage.BucketAttrsToUpdate{VersioningEnabled: true})
```
[the remark this answers is not in the document]
Comment (drakoleat) [reply storage/objects/objects_test.go:62]: sure, will declare a separate variable.

File: storage/objects/list_files_all_version.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+	if err != nil {
+		return fmt.Errorf("storage.NewClient: %v", err)
+	}
+	defer client.Close()
+
+	ctx, cancel := context.WithTimeout(ctx, time.Second*10)
+	defer cancel()
+
+	it := client.Bucket(bucket).Objects(ctx, &storage.Query{
+		// Versions true to output all generations of objects
```
Comment (tritone) [inline storage/objects/list_files_all_version.go:42]: Nit: add period.

File: storage/objects/list_files_all_version.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+	"fmt"
+	"io"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"google.golang.org/api/iterator"
+)
+
+// listFilesAllVersion lists both live and noncurrent versions of objects within specified bucket.
+func listFilesAllVersion(w io.Writer, bucket string) error {
```
Comment (tritone) [inline storage/objects/list_files_all_version.go:29]: I think we should avoid the term "files" for GCS objects if possible (I know other samples already use this, but...). The corresponding Java sample is called `ListObjectsWithOldVersions`, let's use something similar here?

**Comment by tritone** [comment]
CI failure looks like a flake from bigtable. I will re-trigger.

File: storage/objects/list_files_all_version.go (hunk 3 above)
Comment (drakoleat) [reply storage/objects/list_files_all_version.go:42]: added

File: internal/testutil/storage_cleaner.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (drakoleat) [reply internal/testutil/storage_cleaner.go]: refactor to make it easier to understand

**Pull request merged**